src/vm_cpu.rs

```diff
@@ -97,6 +100,195 @@ pub struct VmCpuTrapState {
     pub htval: u64,
     pub htinst: u64,
 }
+#[derive(Default)]
+#[repr(C)]
+struct VmPmuState {
+    // Stores values of PMU CSRs
+    pmu_csrs: [u64; 32],
+    // Stores an index into PmuCounterInfo for configured counters (max index is 31)
+    configured_counters: u64,
+    started_counters: u64,
+}
+
+impl VmPmuState {
+    fn read_pmu_csr(csr: u64) -> Option<u64> {
+        #[inline]
+        fn read_csr<const V: u16>() -> u64 {
+            let r: u64;
+            unsafe {
+                asm!("csrr {rd}, {csr}", rd = out(reg) r, csr = const V);
+            }
+            r
+        }
+        let pmu_info = PmuInfo::get();
+        if Self::is_pmu_csr(csr) && let Some(_) = pmu_info.get_csr_index(csr) {
+            // These represent the range from CSR_CYCLE - CSR_HPMCOUNTER31
+            match csr {
+                0xC00 => Some(read_csr::<0xC00>()),
```

Comment:
I tried a few variations, but it seems like it's difficult to encode in inline assembly without a constant (using in(reg) csr gives an error during compilation). It could just be that my knowledge of inline Rust isn't up to snuff, so a snippet will help.

Reply:
Resolved offline; the CSR operand must be encoded as an immediate value, so short of some macro, there's no easy workaround.

Reply:
Resolved offline; limitation of encoding requires constant values.